score/ScoreBonus.py

```diff
@@ -0,0 +1,275 @@
+from nameinternal import get_card_metadata, get_score_bonus
+from collections import Counter, namedtuple
+from save import Savefile
+
+class ScoreBonus():
+    def __init__(self, name: str):
+        self.name = ""
+        self.format_string = ""
+        self.description = ""
+        self.score_bonus = 0
+
+        if name:
+            data = get_score_bonus(name)
+            self.name = data["NAME"]
+            if data["DESCRIPTIONS"]:
+                self.description = data["DESCRIPTIONS"][0]
+
+    def get_format_string(self) -> str:
+        if self.format_string:
+            return self.format_string
+        else:
+            return f'{self.name}: {self.score_bonus}'
+
+def get_ascension_score_bonus(save: Savefile) -> ScoreBonus:
+    bonus = ScoreBonus("Ascension")
+    if save.ascension_level > 0:
+        # Only applies to the score from the following bonuses:
+        # Floors Climbed
+        # Enemies Killed
+        # Elites Killed
+        # Bosses Slain
+        # Champion
+        # Perfect
+        # Beyond Perfect
+        # Overkill
+        # C-c-c-combo.
+        bonuses_total = 0
+        bonuses_total += get_floors_climbed_bonus(save).score_bonus
+        bonuses_total += get_enemies_killed_bonus(save).score_bonus
+        bonuses_total += get_act1_elites_killed_bonus(save).score_bonus
+        bonuses_total += get_act2_elites_killed_bonus(save).score_bonus
+        bonuses_total += get_act3_elites_killed_bonus(save).score_bonus
+        bonuses_total += get_bosses_slain_bonus(save).score_bonus
+        bonuses_total += get_champions_bonus(save).score_bonus
+        bonuses_total += get_perfect_bosses_bonus(save).score_bonus
+        bonuses_total += get_overkill_bonus(save).score_bonus
+        bonuses_total += get_combo_bonus(save).score_bonus
+        score = round(bonuses_total * 0.05 * save.ascension_level)
+        bonus.format_string = f'Ascension ({save.ascension_level}): {score}'
+        bonus.score_bonus = score
+    return bonus
+
+def get_floors_climbed_bonus(save: Savefile) -> ScoreBonus:
+    bonus = ScoreBonus()
+    bonus.name = "Floors Climbed"
+    bonus.description = "Five points for each Floor you reached"
```

Comment:
Why not pass the name directly to `ScoreBonus`, and use the data from there?

Reply:
because the "name" is passed to the cache and "Floors Climbed" is not in the cache (score_bonuses.json). Unless we can do multiple constructors, this seems correct.